### DISCUSSION TITLE
New major Jolokia protocol version planning (8.0)

### DISCUSSION THREAD

**Discussion by grgrzybek** [rhuss/jolokia#705] [Ideas]
See: https://jolokia.org/reference/html/manual/jolokia_protocol.html#versions

This discussion covers:
* #181
* #250
* #261
* #333
* #564
* #686
* #713
* #726

Tasks
=====

* [x] Bump protocol version to `8.0`
* [x] List operation should allow to filter MBeans for some domains (for example we may not need all 20K MBeans for queues, when broker itself has operations that work on queues) (#261)
* [x] List operation should cache common JSONified MBeanInfo per MBean class (this should be implemented in extensions/plugins, because we may use RBAC data which is different for some queues)
* [x] `org.jolokia.service.jmx.handler.list.DataUpdater()` should be able to delegate to plugin-based extensions which provide additional information (wrap/extend it) to the JSON data of `javax.management.MBeanInfo` (for example RBAC - but this is custom extension, Jolokia shouldn't provide sophisticated extensions, but should handle/cache the results)
* [x] JSONified `javax.management.MBeanInfo` should also contain `keys` field with a map of key-value pairs extracted from the ObjectName - that'd be handy for clients that construct a JMX tree (cc: @tadayosi)
* List operation should allow for pagination (for lazy loading) and handle dynamic MBean (un)registration in-between pagination requests (this may be hard to achieve) - I'm removing pagination from the scope of this discussion. Decreasing the number of MBeans can be achieved using list filtering and pagination should be done at higher level (for example in Hawtio). Jolokia provides JMX notification support and we can register a listener for `JMImplementation:type=MBeanServerDelegate` and `JMX.mbean.registered` event.

Description
=====

I'm working on performance related issues and my playground is an Artemis (2.35.0) broker with 10,000 queues created. For each queue we have two MBeans, for example for queue named `grgr-q-000942` we have these MBean names
* `broker=\"0.0.0.0\",component=addresses,address=\"grgr-q-000942\"`
* `broker=\"0.0.0.0\",component=addresses,address=\"grgr-q-000942\",subcomponent=queues,routing-type=\"anycast\",queue=\"grgr-q-000942\"`

With 10,000 queues, `{"type":"list"}` request takes 15 seconds and the response is 234MB in size.

The problem is that for each queue, there are two distinct MBeanInfos being JSONified 10000 times containing operations and attributes of an MBean class of Artemis queue...

Fix for Hawtio https://issues.redhat.com/browse/ENTESB-4055 contains an optimized version of Jolokia's `list()` operation and [Hawtio](https://github.com/hawtio/hawtio-next) decides between calling Jolokia `list()` or special operation on something called _RBAC Registry_, which effectively caches the MBeanInfo data and simply uses something like this:

```json
"broker=\"0.0.0.0\",component=addresses,address=\"grgr-q-000359\"": "activemq.artemis:address"
```

meaning that this queue's MBeanInfo is stored under `activemq.artemis:address` key in the cache.

This was real life-saver, but it'd be nice to have this optimization in Jolokia itself. When running Jolokia using `-javaagent`, there's no Hawtio acting as a proxy (and providing own MBeans) and we can't rely on external optimizations.

cc: @tadayosi

**Comment by tadayosi** [comment]
That's a great idea. Just a few questions on the planned tasks.

> Emphasize that processing parameters should be sent with config field of the request payload instead of query parameters of the request (that's a cosmetic issue)

Why do we need to enforce this?  I agree that the optimised JSON output should be optional by either config field or query parameter though.

> List operation should allow to filter MBeans for some domains (for example we may not need all 20K MBeans for queues, when broker itself has operations that work on queues)

I thought the current List op already supports the feature.

**Reply by grgrzybek** [reply]
> That's a great idea. Just a few questions on the planned tasks.
> 
> > Emphasize that processing parameters should be sent with config field of the request payload instead of query parameters of the request (that's a cosmetic issue)
> 
> Why do we need to enforce this? I agree that the optimised JSON output should be optional by either config field or query parameter though.

Maybe I didn't express it correctly. I rather wanted to review how it's handled and whether we could add some additional parameters (like paging). But this is not that important.

> 
> > List operation should allow to filter MBeans for some domains (for example we may not need all 20K MBeans for queues, when broker itself has operations that work on queues)
> 
> I thought the current List op already supports the feature.

True - with `"path":"java.lang/type=Runtime"` we can get something like:

```json
{
  "request": {
    "path": "java.lang/type=Runtime",
    "type": "list"
  },
  "value": {
    "attr": {
      "SpecVendor": {
        "rw": false,
        "type": "java.lang.String",
        "desc": "SpecVendor"
      },
...
```

But what if you want _everything but all the queues_? I'll think about this.

**Comment by fbacchella** [comment]
It might be interesting to be able to plug in different serialization/deserialization protocols. JSON is very inefficient and error-prone. Formats like MessagePack or CBOR could be more appropriate.

I don’t think Protobuf would be a good choice. It’s very efficient for a closed space of possible responses, but for something very generic like MBean serialization, it doesn't fit.

**Comment by grgrzybek** [comment]
Thanks for hinting me about [MessagePack](https://msgpack.org/) and [CBOR](https://cbor.io/).
This would be big in Jolokia to make serialization protocol pluggable ;) But still we may consider it.

Although:
* https://www.npmjs.com/package/msgpack-lite looks pretty dead
* https://www.rfc-editor.org/rfc/rfc8949.html (CBOR) sounds dangerously similar to ASN.1's DER/BER encodings
* https://cbor.io/impls.html mentioning `bower install cbor` also sounds dangerous (https://github.com/paroga/cbor-js is  dead too)

And JSON is simply too ubiquitous ;) And yes - I'm a fan of XML too - it's verbose, but does it's job perfectly where suitable ;)

**Reply by fbacchella** [reply]
Indeed you have complex needs. I’m using Jolokia in application that already embed jackson. So a plugable protocol would be easy for me to use. If you want to be compatible with javascript application, JSON is a no brainer.

**Reply by fbacchella** [reply]
DER/BER is not that bad. It’s all the ceremonial with ASN.1 grammars that is a pain in the neck. I think DER/BER without schema and OID is possible.

**Reply by grgrzybek** [reply]
> DER/BER is not that bad. It’s all the ceremonial with ASN.1 grammars that is a pain in the neck. I think DER/BER without schema and OID is possible.

don't get me wrong - DER/BER is amazing! that's huge part of my professional career back in ~2002.

**Comment by grgrzybek** [comment]
I'm closing the discussion. Further improvements can be made under separate issues or discussions.
